crates/turborepo-scm/src/hash_object.rs

```diff
@@ -0,0 +1,192 @@
+use std::{
+    io::{BufWriter, Read, Write},
+    panic,
+    process::{Command, Stdio},
+    thread,
+};
+
+use nom::{Finish, IResult};
+use turbopath::{AbsoluteSystemPathBuf, RelativeUnixPathBuf};
+
+use crate::{package_deps::GitHashes, Error};
+
+pub(crate) fn hash_objects(
+    pkg_path: &AbsoluteSystemPathBuf,
+    to_hash: Vec<RelativeUnixPathBuf>,
+    pkg_prefix: &RelativeUnixPathBuf,
+    hashes: &mut GitHashes,
+) -> Result<(), Error> {
+    if to_hash.is_empty() {
+        return Ok(());
+    }
+    let mut git = Command::new("git")
+        .args(["hash-object", "--stdin-paths"])
+        .current_dir(pkg_path)
+        .stdout(Stdio::piped())
+        .stderr(Stdio::piped())
+        .stdin(Stdio::piped())
+        .spawn()?;
+    {
+        let stdout = git
+            .stdout
+            .as_mut()
+            .ok_or_else(|| Error::git_error("failed to get stdout for git hash-object"))?;
+        // We take, rather than borrow, stdin so that we can drop it and force the
+        // underlying file descriptor to close, signalling the end of input.
+        let stdin: std::process::ChildStdin = git
+            .stdin
+            .take()
+            .ok_or_else(|| Error::git_error("failed to get stdin for git hash-object"))?;
+        let mut stderr = git
+            .stderr
+            .take()
+            .ok_or_else(|| Error::git_error("failed to get stderr for git hash-object"))?;
+        let result = read_object_hashes(stdout, stdin, &to_hash, pkg_prefix, hashes);
+        if let Err(err) = result {
+            let mut buf = String::new();
+            let bytes_read = stderr.read_to_string(&mut buf)?;
+            if bytes_read > 0 {
+                // something failed with git, report that error
+                return Err(Error::git_error(buf));
+            }
+            return Err(err);
+        }
+    }
+    git.wait()?;
+    Ok(())
+}
+
+const HASH_LEN: usize = 40;
+
+fn read_object_hashes<R: Read, W: Write + Send>(
+    mut reader: R,
+    writer: W,
+    to_hash: &Vec<RelativeUnixPathBuf>,
+    pkg_prefix: &RelativeUnixPathBuf,
+    hashes: &mut GitHashes,
+) -> Result<(), Error> {
+    thread::scope(move |scope| -> Result<(), Error> {
+        let write_thread = scope.spawn(move || -> Result<(), Error> {
+            let mut writer = BufWriter::new(writer);
+            for path in to_hash {
+                path.write_escaped_bytes(&mut writer)?;
+                writer.write_all(&[b'\n'])?;
+                writer.flush()?;
+            }
+            // writer is dropped here, closing stdin
+            Ok(())
+        });
+        // Buffer size is HASH_LEN + 1 to account for the trailing \n
+        let mut buffer: [u8; HASH_LEN + 1] = [0; HASH_LEN + 1];
+        for (i, filename) in to_hash.iter().enumerate() {
+            if i == to_hash.len() {
```

Comment:
I'm pretty sure that this doesn't do a safety check on the reader being entirely consumed at the same time as the file list in `to_hash` is consumed.

`write_escaped_bytes` _should_ prevent this from being an issue, but the test for newlines didn't get ported to Rust so I don't know if it works:
https://github.com/vercel/turbo/blob/main/cli/internal/hashing/package_deps_hash_test.go#L45-L115

I'm less certain about the behavior of a broken pipe on inputs happening on the last path in the list.